src/main/java/com/microsoft/azure/vmagent/AzureVMCloud.java

```diff
@@ -232,12 +248,88 @@ public String getAzureCredentialsId() {
     }
 
     /**
+     * Ensures there is a valid template list available.
+     * Must be called under the vmTemplatesListLock
+     */
+    private void ensureVmTemplateList() {
+        if (vmTemplates == null) {
+            vmTemplates = new ArrayList<AzureVMAgentTemplate>();
+        }
+    }
+
+    /**
+     * Remove all templates in the template set.
+     */
+    public final void clearVmTemplates() {
+        // Obtain lock while we copy the list.
+        synchronized (this) {
+            ensureVmTemplateList();
+            vmTemplates.clear();
+        }
+    }
+
+    /**
+     * Adds a new VM template to the store
+     */
+    public final void addVmTemplate(AzureVMAgentTemplate newTemplate) {
+        // Obtain lock while we copy the list.
+        synchronized (this) {
+            ensureVmTemplateList();
+            vmTemplates.add(newTemplate);
+            newTemplate.setAzureCloud(this);
+        }
+    }
+
+    /**
+     * Sets the template list to a new template list.  The list is cleared and
+     * the elements are iteratively added (to avoid leakage of the
+     * internal template list)
+     * @param newTemplates Template set to use
+     */
+    public final void setVmTemplates(List<AzureVMAgentTemplate> newTemplates) {
+        synchronized (this) {
+            ensureVmTemplateList();
+            vmTemplates.clear();
+            for (AzureVMAgentTemplate newTemplate : newTemplates) {
+                vmTemplates.add(newTemplate);
+                newTemplate.setAzureCloud(this);
+            }
+        }
+    }
+
+    /**
+     * Removes a template from the template store by name.
+     * Removes all templates with the template name.
+     * @param templateName Template name to remove
+     */
+    public final void removeVmTemplate(String templateName) {
+        // Obtain lock while we copy the list.
+        synchronized (this) {
+            ensureVmTemplateList();
+            // in 1.8 we can use lambdas here, but in order to maintain
+            // 1.7 compatibility we don't.
+            ArrayList<AzureVMAgentTemplate> removeTemplates = new ArrayList<AzureVMAgentTemplate>();
+            for (AzureVMAgentTemplate template : vmTemplates) {
+                if (template.getTemplateName().equals(templateName)) {
```

Comment:
dedupe only for backwards compatibility. Afterwards it shouldn't happend, you don't add a template if the name is already taken